Stats Stat!

danwashable opened the pull request:
Resolves ceros/ceros-product#1385
Resolves ceros/ceros-product#1374

Product PR: ceros/ceros-product#1386
Chef PR: ceros/ceros-product-chef#111

noelcrowdfusion commented on src/server/stats.js:
```diff
@@ -0,0 +1,131 @@
+        submittedOps.unshift(currentTime);
+
+        // We want to prevent the tracking array from getting too large, so if the last element is over a minute old, remove it
+        var oneMinuteAgo = currentTime - 1000;
```
So this'll only pop off one item if it's old?  That means the array will always have the same number of items as however many ops were submitted within the most active minute?  That's probably ok but it seems like purging all items that are over a minute old is more what we want.

danwashable replied:
So I did it this way so that if there's a large number of ops, it won't slow down the submitting and yet it will still keep the array from getting too large. I have no idea how many ops will come in over a minute and it'd suck to loop through thousands just to submit a single op. That being said, this is actually just a fallback in the case that the collectd polling isn't happening anymore because the polling does truncating.

noelcrowdfusion commented on src/server/stats.js:
```diff
@@ -0,0 +1,131 @@
+
+    /**
+     * Get the number of submitted ops. If interval is not specified, get all. If interval is specified, return
+     * the number within that amount of time and then truncate older values off.
```
Do we want to report just the number of ops within the interval or the average ops within that interval over some greater interval.  So like, let's say we want ops per second, we have to poll exactly every second; if we miss a second we'll lose that data for that second.  Also, the graph will probably be more jumpy and harder to look at.

danwashable replied:
I wasn't really sure on this. Doing it over an interval means that you potentially lose spikes in the data. If a large number all come in at the same time, it averages it out and makes it harder to see.

noelcrowdfusion commented on src/server/stats.js:
```diff
@@ -0,0 +1,131 @@
+
+            for (var i = 0; i < submittedOps.length; i++) {
+                if (submittedOps[i] < intervalStart) {
+                    submittedOps.length = i;
```
Does manually setting `length` on an array automatically truncate it?  Is that better than doing a `splice`?

danwashable replied:
It does and it's actually faster than `splice`!

noelcrowdfusion commented on src/server/stats.js:
```diff
@@ -0,0 +1,131 @@
+    },
+
+    /**
+     * An event was just broadcast. Add a new timestamp to the tracking array. Add it at the beginning so
```
Same questions here.

danwashable replied:
Same answers here.

noelcrowdfusion commented:
Code review complete.  Some questions about the intervals, otherwise good.

danwashable commented:
Ok @noelcrowdfusion. Addressed code review as we discussed. I'm going to look into the failing build. It's been failing since a previous release but it'd be nice to fix it.

danwashable commented:
Welp, got one error fixed by npm installing coffee-script (my vm feels tainted now) but now I'm running into some sort of db error. Probably not worth the time at the moment to fix this.

noelcrowdfusion commented on src/server/stats.js:
```diff
@@ -0,0 +1,145 @@
+        return {
+            openConnectionCount : this.getUserAgentCount(),
+            openDocCount        : this.getOpenDocCount(),
+            submittedOpsCount   : submittedOpCount,
```
naming: it's not really a count if it's per second.  I know this probably means changing the chef code but I think it's worth it to not be confusing.

danwashable replied:
Word. Changing chef code ain't so bad now that hubot refetches!

noelcrowdfusion commented on src/server/stats.js:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @returns {Number}
+     */
+    getSubmittedOpCount : function() {
```
More naming naziness: call this flush submitted ops or something so that it's clear that it resets the ops?

noelcrowdfusion commented on src/server/stats.js:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @returns {Number}
+     */
+    getBroadcastEventCount : function() {
```
Same thing here with the naming

noelcrowdfusion commented:
Coupla comments on naming.  Otherwise looks good.

danwashable commented:
Ok @noelcrowdfusion, did the renamings!

noelcrowdfusion commented:
Code :+1: :1234: :toilet:
